apps/service_providers/llm_service/runnables.py

```diff
@@ -267,25 +270,95 @@ def invoke(
 
         if config.get("configurable", {}).get("save_input_to_history", True):
             chat_message = self.state.save_message_to_history(input, ChatMessageType.HUMAN)
-            if resource_file_ids:
-                chat_message.metadata = chat_message.metadata | resource_file_ids
-                chat_message.save()
+            if human_message_resource_file_ids:
+                self._add_resource_files_to_message_metadata(chat_message, human_message_resource_file_ids)
 
         # Note: if this is not a new chat then the history won't be persisted to the thread
         thread_id = self.state.get_metadata(Chat.MetadataKeys.OPENAI_THREAD_ID)
         if thread_id:
             input_dict["thread_id"] = thread_id
 
         input_dict["instructions"] = self.state.get_assistant_instructions()
-
         response = self._get_response_with_retries(config, input_dict, thread_id)
+
+        output, ai_message_resource_file_ids = self._save_response_annotations(response)
+
         if not thread_id:
             self.state.set_metadata(Chat.MetadataKeys.OPENAI_THREAD_ID, response.thread_id)
 
+        ai_message = self.state.save_message_to_history(output, ChatMessageType.AI)
+        if ai_message_resource_file_ids:
+            self._add_resource_files_to_message_metadata(ai_message, ai_message_resource_file_ids)
+
+        return ChainOutput(output=output, prompt_tokens=0, completion_tokens=0)
+
+    @transaction.atomic()
+    def _save_response_annotations(self, response: OpenAIAssistantFinish) -> tuple[str, dict]:
+        """
+        This makes a call to OpenAI with the `run_id` and `thread_id` to get more information about the response
+        message, specifically regarding annotations.
+        - Those of type `file_citation` cannot be read, but since we already have those files, we should be fine
+        by only storing the reference to the file i.e. its external_id
+        - Those of type `file_path` are generated and can be downloaded. A file is created in OCS for each of these
+
+        """
+        from apps.assistants.sync import get_and_store_openai_file
+
+        client = self.state.get_openai_assistant().client
+        page: SyncCursorPage = client.beta.threads.messages.list(response.thread_id, run_id=response.run_id)
+        # We use a run to get a response, so it's safe to assume a run will have a single message
+        message: Message = page.data[0]
+        message_content = message.content[0]
+        annotations = message_content.text.annotations
         output = response.return_values["output"]
-        self.state.save_message_to_history(output, ChatMessageType.AI)
+        resource_file_ids = {}
```

Comment:
nit: using a defaultdict would mean you don't have to check if the key is present
```suggestion
        resource_file_ids = defaultdict(list)
```

```python
 if annotation.type not in resource_file_ids:
    resource_file_ids[annotation.type] = [file_id]
else:
    resource_file_ids[annotation.type].append(file_id)
```

becomes
```python
resource_file_ids[annotation.type].append(file_id)
```

Reply:
This is cool, thanks!